### ISSUE TITLE
Calculate Full Team@Event Status Serverside

### ISSUE THREAD

**Issue by phil-lopreiato** [the-blue-alliance/the-blue-alliance#1628]
As we use this more and more around the site, we should not be calculating this data both server-side and in the app. Some examples for testing can be found on [this fb thread](https://www.facebook.com/groups/moardata/permalink/1206221399398153/)

Here's what I'm thinking the json result should look like:

```
{
  'rank': {
    'rank': 1,
    'total': 60,
    'played': 12,
    'first_sort': 102 (first order sort item)
    'record': "12-0-0",
    'breakdown': string of all the ranking data for this team, like what shows in the app now
  },
  'alliance': {
    'position': one of {captain, 1st, 2nd, 3rd, backup},
    'name': name of alliance ("Alliance #1" or "Newton"),
    'backup': "backup" dict from alliance row
  },
  'playoff': {
    'status': one of {playing, eliminated, won (the event)},
    'round': one of {ef, qf, sf, f},
    'record': record in the playoffs
    'current_level_record': the team's record in the most recent playoff level (only if status is playing)
  }
}
```

**Comment by bdaroz** [comment]
You'd make my weekend with this. The number of calls we have to make for the Slack Bot to calculate this....

Is this for a new team & event specific API endpoint, or is this an extension/modification of `/event/xxxxx/rankings`? If it's the latter we need to identify that team that is in the `rank`, and `alliance` and `playoff` objects should be children of `rank`, so we only iterate over 1 parent object.

We should probably also clean up the W-L-T qualification only or event-wide kerfuffle (or include a 2nd w/ elims included).

**Comment by phil-lopreiato** [comment]
> Is this for a new team & event specific API endpoint, or is this an extension/modification of /event/xxxxx/rankings?

The former. This will be an apiv3 thing.

> We should probably also clean up the W-L-T qualification only or event-wide kerfuffle (or include a 2nd w/ elims included).

It'll be qual record. That's what we can get from the ranking data FIRST provides and not have to query/traverse all matches.

**Comment by bdaroz** [comment]
> It'll be qual record. That's what we can get from the ranking data FIRST provides and not have to query/traverse all matches.

Which leads us to the other discussions on W-L-T not matching the TBA website, or the played matches. (See also #1464 especially the Apr 12 comment, and the-blue-alliance/the-blue-alliance-android#729)

I would argue that the API should be able to produce the same data presented on the website, and further, users should expect to see matching data on the website and from an API-using app. TBA already has the functions to calculate that W-L-T, albeit by walking the matches, but that's less intensive than doing the same _plus_ wrapping/unwrapping API calls doing it.

**Comment by phil-lopreiato** [comment]
That's the exact motivation for doing this. The app and site currently calculate things in different ways - if we do it in one place, then all the sources can have the same data to display.

Maybe I don't understand your question?

**Comment by bdaroz** [comment]
Let me take another stab at this:

The webpages right now, lets say [Team 987's page](https://www.thebluealliance.com/team/987) for a concrete example. If you look at the LA Regional, the webpage lists 12-3-0 for W-L-T. If you poll the API for `/api/v2/event/2016calb/rankings` it shows them 6-2-0 at rank 6.

As best I can can tell from the `blame` the website has been calculating its own W-L-T since 29102404cef4c6f3a311350bfd6a4f45db44f015 landed in Feb 2011. That calculation is across all matches, both quals and elims.

> It'll be qual record. That's what we can get from the ranking data FIRST provides and not have to query/traverse all matches.

My argument is to standardize and provide the same calculation, __across all matches__ that the webpage provides, via the API. __Not__ to only provide the _FIRST_ data, which only covers quals.

In other words, W-L-T via the API should == W-L-T from the HTML UI. I also don't think "fixing" the HTML UI to use the _FIRST_ data (and drop elims from the W-L-T) would go over well, especially when apps are walking the matches now via the API specifically because the TBA HTML UI != API.

**Comment by phil-lopreiato** [comment]
Gotcha, thanks for clarifying your idea. 

Now that you point it out, I actually think the copy text on the team page is what's wrong here. That page implies that the record (including all matches) is tied to the team's rank, which is untrue - rank and *qual record* are what typically go hand in hand. I actually think we nailed [in 2015](https://www.thebluealliance.com/team/987/2015), since we draw a distinction between quals and playoffs. 

I would actually propose that we change those text blurbs to be a more complete picture of the event, maybe something like:
 - **Rank 2** with **14 RP** and qual record of **10-2-0**
 - **Captain** of **Alliance 2**
 - Knocked out in the **semifinals**
 - Won the the following awards: (as is currently)

We can build all of that from the data I have outlined above (which means more code sharing internal to TBA, as well), and I think it gives a much better overview of how the team performed at the event. 

Now, there's still two questions that I don't know the answer to (or even if they're things we should care about):
 - How much does it matter that we don't provide the record in elims for a team? I am of the opinion that "Knocked out in the semis" is much more useful than "Had a record of 5-2 in playoffs". But this one should definitely be open to discussion.
 - We still display a team's overall record in a year on the web (separated by official and offseason events). I think that's okay to include playoffs, since it's an aggregate performance, rather than details at a specific event. Although we still aren't exposing that via the API, and probably won't unless we can think of a good reason to.

**Comment by bdaroz** [comment]
Thanks Phil - I have no problem making that change as I think it provides some more useful information overall, I'd just advocate for the same information being readily available in the API call as well.

To your questions:
> How much does it matter that we don't provide the record in elims for a team?

I agree a 5-2 record can be ambiguous w/o the "Knocked out..." text to go with it. But knowing they are eliminated in "x" round explains 2 losses... but were the earlier rounds 2-1 or 2-0? There's a debatable amount of value to knowing that, but I think the value would outweigh the computational cost.

Perhaps "Knocked out in the semifinals with a record of 3-2-0" is the HTML UI compromise, and the API adds a separate elim W-L-T. (This would also allow the API client to sum the qual+elim values if so desired.)

That also addresses...

>We still display a team's overall record in a year on the web (separated by official and offseason events). I think that's okay to include playoffs, `...` Although we still aren't exposing that via the API, and probably won't unless we can think of a good reason to.

I tend to try to think of what to include or not include as a relation to value vs computational cost. I don't think, for this one, providing that aggregate total on a `team/year` endpoint is worth the cost. Clients that need that would be able to, under my proposal above, request the individual attended event rankings to get the component pieces to create the total. I think it's quite possible that the overall W-L-T would not be the only reason a client would make those `event/ranking` calls.

**Comment by phil-lopreiato** [comment]
> Perhaps "Knocked out in the semifinals with a record of 3-2-0" is the HTML UI compromise, and the API adds a separate elim W-L-T. (This would also allow the API client to sum the qual+elim values if so desired.)

I like this idea, let's do it!
